apps/prairielearn/src/migrations/20250707220720_enrollments__meta_columns__add.sql

```diff
@@ -0,0 +1,110 @@
+CREATE TYPE enum_enrollment_status AS ENUM(
+  'invited',
+  'joined',
+  'removed',
+  'rejected',
+  'blocked',
+  'lti13_pending'
+);
+
+ALTER TABLE enrollments
+ADD COLUMN status enum_enrollment_status NOT NULL DEFAULT 'joined';
+
+ALTER TABLE enrollments
+ADD COLUMN lti_managed BOOLEAN NOT NULL DEFAULT FALSE;
+
+ALTER TABLE enrollments
+ADD COLUMN pending_uid TEXT;
+
+ALTER TABLE enrollments
+ADD COLUMN pending_lti13_sub TEXT;
+
+-- If a lti13_course_instance is deleted, we want to delete the enrollment.
+-- This should only happen if the user is in the 'lti13_pending' state.
+ALTER TABLE enrollments
+ADD COLUMN pending_lti13_instance_id BIGINT REFERENCES lti13_course_instances (id) ON DELETE CASCADE ON UPDATE CASCADE;
+
+-- lti13_course_instances can be hard deleted, which cascades to deleting associated enrollments.
+-- We index this column to speed up the deletion of those enrollments and avoid a table scan.
+CREATE INDEX enrollments_pending_lti13_instance_id_idx ON enrollments (pending_lti13_instance_id);
+
+ALTER TABLE enrollments
+ALTER COLUMN user_id
+DROP NOT NULL;
+
+-- Forbid impossible states: rejected + lti_synced, blocked + lti_synced
+ALTER TABLE enrollments
+ADD CONSTRAINT enrollments_impossible_synced CHECK (
+  NOT (
+    status IN ('rejected', 'blocked')
+    AND lti_managed
+  )
+);
+
+-- If a user is invited or rejected an invitation, we don't link them to a user to avoid PII leakage.
+-- A user in any other state must have a user_id.
+-- Further discussion: https://github.com/PrairieLearn/PrairieLearn/issues/12198#issuecomment-3161792357
+ALTER TABLE enrollments
+ADD CONSTRAINT enrollments_user_id_null_only_if_invited_rejected CHECK (
+  (status IN ('invited', 'rejected')) = (user_id IS NULL)
+);
+
+-- Only users in the 'invited' state can have a pending_uid.
+ALTER TABLE enrollments
+ADD CONSTRAINT enrollments_pending_uid_null_only_if_invited CHECK ((status = 'invited') = (pending_uid IS NOT NULL));
```

Comment:
I think this constraint is correct, but it contradicts the earlier one. See https://github.com/PrairieLearn/PrairieLearn/pull/12363/files#r2280178732 and https://github.com/PrairieLearn/PrairieLearn/pull/12363/files#r2280183714